File: src-rb/renderer.rb
require "js"

module Kotoyomi
  module Renderer
    def self.render(cues, container)
      document = JS.global[:document]
      container[:innerHTML] = ""

      elements = JS.eval("return [];")
      cue_count = cues[:length].to_i
      cue_count.times do |i|
        cue = cues[i]
        div = document.createElement("div")
        cue_id = cue[:id].to_s
        div[:id] = cue_id.empty? ? "stanza-#{i + 1}" : cue_id
        div[:className] = "stanza"
        div[:dataset][:startTime] = cue[:startTime].to_s

        cue[:text].to_s.split("\n").each do |line|
          p = document.createElement("p")
          p[:className] = "stanza-line"
          p[:textContent] = line
          div.appendChild(p)
        end

        container.appendChild(div)
        elements.push(div)
      end
      elements
    end
  end
end

File: src/renderer.ts
import type { PoemStanza } from "./types.ts";

export function renderPoem(stanzas: PoemStanza[], container: HTMLElement): HTMLElement[] {
  container.replaceChildren();

  const elements: HTMLElement[] = [];
  for (const stanza of stanzas) {
    const div = document.createElement("div");
    div.id = stanza.id;
    div.className = "stanza";
    div.dataset.time = String(stanza.time);

    for (const line of stanza.lines) {
      const p = document.createElement("p");
      p.className = "stanza-line";
      p.textContent = line;
      div.appendChild(p);
    }

    container.appendChild(div);
    elements.push(div);
  }
  return elements;
}
